src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+
+    /// This method should set the number of threads to use.
+    fn set_nthreads_(&mut self, n: usize);
+
+    // *********************************************************** //
+    // ******************** PROVIDED METHODS ********************* //
+    // *********************************************************** //
+
+    fn collapse_<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: io::BufRead,
+        W: io::Write,
+    {
+        let mut occurrences = Occurrences::new(self.nthreads());
+
+        // Consume the header, if any, and do any other pre-processing
+        // that needs to occur.
+        self.pre_process(&mut reader, &mut occurrences)?;
+
+        // Do collapsing.
+        if occurrences.is_concurrent() {
+            self.collapse_multi_threaded(reader, &mut occurrences)?;
+        } else {
+            self.collapse_single_threaded(reader, &mut occurrences)?;
+        }
+
+        // Write results.
+        occurrences.write_and_clear(writer)
+    }
+
+    fn collapse_multi_threaded<R>(
+        &mut self,
+        mut reader: R,
+        occurrences: &mut Occurrences,
+    ) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let nstacks_per_job = self.nstacks_per_job();
+        let nthreads = self.nthreads();
+
+        assert_ne!(nstacks_per_job, 0);
+        assert!(nthreads > 1);
+        assert!(occurrences.is_concurrent());
+
+        crossbeam::thread::scope(|scope| {
+            // Channel for sending an error from the worker threads to the main thread
+            // in the event a worker has failed.
+            let (tx_error, rx_error) = channel::bounded::<io::Error>(1);
+
+            // Channel for sending input data from the main thread to the worker threads.
+            let (tx_input, rx_input) = channel::bounded::<Option<Vec<u8>>>(2 * nthreads);
+
+            // Channel for worker threads that have errored to signal to all the other
+            // worker threads that they should stop work immediately and return.
+            let (tx_stop, rx_stop) = channel::bounded::<()>(nthreads - 1);
+
+            let mut handles = Vec::with_capacity(nthreads);
+            for _ in 0..nthreads {
+                let tx_error = tx_error.clone();
+                let rx_input = rx_input.clone();
+                let (tx_stop, rx_stop) = (tx_stop.clone(), rx_stop.clone());
+
+                let mut folder = self.clone();
+                let mut occurrences = occurrences.clone();
+
+                // Launch the worker thread...
+                let handle = scope.spawn(move |_| loop {
+                    channel::select! {
+                        recv(rx_input) -> input => {
+                            // Received from the main thread either:
+                            // * `Some(<input_data>)`, or
+                            // * `None` = a signal that no more input data will be sent.
+                            let data = match input.unwrap() {
```

Comment:
I don't think you need the `Option<Vec<u8>>` here. Instead, `recv` will return `Err` when `rx_input` is closed (i.e., when all senders have been dropped). So you can instead just use `match input` and then return in the `Err` case.